### PR TITLE
VACMS-3494: Add non-beneficiaries to r&s tag selections.

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
See #3494 

## Testing done
Visual / Behat

## Screenshots
image

## QA steps
- As an administrator, go to `/node/9544/edit`
- Visually verify that non-beneficiaries are available in the Audience dropdown
- Visually verify that when either one Beneficiary, or non-beneficiary item is selected, no more than 3 topics can be selected.

### Definition of Done
- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Core Application Team`
- [x] `Product Support Team`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change. 
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

**Review by acrollet — CHANGES_REQUESTED** [review]
Ok, I just found an issue that I think was probably there in my original code. To reproduce, go to http://pr3615.ci.cms.va.gov/node/9544/edit and:

- Select 'none' for Audience
- Choose beneficiaries or non-beneficiaries
- Check all 4 topics tags

The beneficiaries select dropdown is correctly disabled, but it's still possible to select one of the beneficiaries radio buttons.

One more issue:

- Check two or fewer topics tags
- Select "Beneficiaries" and choose a beneficiary tag
- Select "Non-beneficiaries" and choose a non-beneficiary tag
- Save the node

Both the Beneficiaries and Non-beneficiaries values are stored on the node. I think we should also enhance the validation in [_va_gov_backend_audience_topics_validation](https://github.com/ethanteague/va.gov-cms/blob/44b95850c89b4963cbb794328495d8132bf22913/docroot/modules/custom/va_gov_backend/va_gov_backend.module#L523) to ensure that terms from both vocabularies can't be selected.

**Comment by ethanteague** [comment]
> Ok, I just found an issue that I think was probably there in my original code. To reproduce, go to http://pr3615.ci.cms.va.gov/node/9544/edit and:
> 
> * Select 'none' for Audience
> * Choose beneficiaries or non-beneficiaries
> * Check all 4 topics tags
> 
> The beneficiaries select dropdown is correctly disabled, but it's still possible to select one of the beneficiaries radio buttons.
> 
I don't think this is an issue - once none is selected, logic changes both vocab selections to the N/A option. So, if you have one of the radios selected in Beneficiaries, and then toggle the selector to none, and then back to Beneficiaries, you will see that the radio toggle that was previously set is now unset.

> One more issue:
> 
> * Check two or fewer topics tags
> * Select "Beneficiaries" and choose a beneficiary tag
> * Select "Non-beneficiaries" and choose a non-beneficiary tag
> * Save the node
> 
> Both the Beneficiaries and Non-beneficiaries values are stored on the node. I think we should also enhance the validation in [_va_gov_backend_audience_topics_validation](https://github.com/ethanteague/va.gov-cms/blob/44b95850c89b4963cbb794328495d8132bf22913/docroot/modules/custom/va_gov_backend/va_gov_backend.module#L523) to ensure that terms from both vocabularies can't be selected.
:+1:

**Comment by acrollet** [comment]
argh, devshop lost the PR environment 🤦‍♂️

**Comment by acrollet** [comment]
>> Select 'none' for Audience
>> Choose beneficiaries or non-beneficiaries
>> Check all 4 topics tags
>> The beneficiaries select dropdown is correctly disabled, but it's still possible to select one of the beneficiaries radio buttons.

> I don't think this is an issue - once none is selected, logic changes both vocab selections to the N/A option. So, if you have one of the radios selected in Beneficiaries, and then toggle the selector to none, and then back to Beneficiaries, you will see that the radio toggle that was previously set is now unset.

Hmm, I do think it's an issue, because it's possible to select a total of 5 terms when following those steps. That said, the form validation still catches the problem and doesn't allow saving the node in that state. @oksana-c do you think we should fix this now, make a follow-on, or not worry about it?

**Comment by acrollet** [comment]
I validated the fix, I can't select terms from both beneficiaries and non-beneficiaries now. 👍

**Comment by acrollet** [comment]
Couple of new things I ran across:

**Thing 1:**

Screen Shot 2020-12-04 at 7 24 52 AM

**Thing 2:**

Steps to repro:
- Select a Non-beneficiary tag
- Save the node and confirm that the tag was stored
- Edit the node

Non-beneficiaries is selected in the 'Audience' dropdown, but the Beneficiaries selection is shown and no radio is selected.

**Comment by ethanteague** [comment]
> > > Select 'none' for Audience
> > > Choose beneficiaries or non-beneficiaries
> > > Check all 4 topics tags
> > > The beneficiaries select dropdown is correctly disabled, but it's still possible to select one of the beneficiaries radio buttons.
> 
> > I don't think this is an issue - once none is selected, logic changes both vocab selections to the N/A option. So, if you have one of the radios selected in Beneficiaries, and then toggle the selector to none, and then back to Beneficiaries, you will see that the radio toggle that was previously set is now unset.
> 
> Hmm, I do think it's an issue, because it's possible to select a total of 5 terms when following those steps. That said, the form validation still catches the problem and doesn't allow saving the node in that state. @oksana-c do you think we should fix this now, make a follow-on, or not worry about it?

My bad - I see it, @acrollet. Added a fix.

**Comment by ethanteague** [comment]
> Couple of new things I ran across:
> 
> **Thing 1:**
> 
> Screen Shot 2020-12-04 at 7 24 52 AM
> 
> **Thing 2:**
> 
> Steps to repro:
> 
> * Select a Non-beneficiary tag
> * Save the node and confirm that the tag was stored
> * Edit the node
> 
> Non-beneficiaries is selected in the 'Audience' dropdown, but the Beneficiaries selection is shown and no radio is selected.

:+1: - fixed, @acrollet

**Comment by acrollet** [comment]
@ethanteague sorry man, this stuff is like whack-a-mole! New issue:

- Ensure that a beneficiary term is selected
- Select 'None' for Audience
- Select 4 topic terms
- Unselect two topic terms
- Select 'Non-beneficiaries' for Audience
- Select a Non-beneficiary term
- Select 'None' for Audience

Only one topic term can be selected. (Only 3 rather than 4)

Previous issues:

- proofing view is fixed 👍 
- non-beneficiaries default value is present on the form 👍

**Comment by ethanteague** [comment]
> @ethanteague sorry man, this stuff is like whack-a-mole! New issue:
> 
> * Ensure that a beneficiary term is selected
> * Select 'None' for Audience
> * Select 4 topic terms
> * Unselect two topic terms
> * Select 'Non-beneficiaries' for Audience
> * Select a Non-beneficiary term
> * Select 'None' for Audience
> 
> Only one topic term can be selected. (Only 3 rather than 4)
> 
> **Previous issues:**
> 
> * proofing view is fixed +1
> * non-beneficiaries default value is present on the form +1

Apologies for missing that, @acrollet - appreciate you taking the time to make sure this thing is working correctly. :+1:

**Comment by ethanteague** [comment]
Gonna have to screenshare with you on Monday, @acrollet - I'm not able to break it like so:
Ensure that a beneficiary term is selected

- Select 'None' for Audience
- Select 4 topic terms
- Unselect two topic terms
- Select 'Non-beneficiaries' for Audience
- Select a Non-beneficiary term
- Select 'None' for Audience

**Comment by ethanteague** [comment]
Alrighty, @acrollet - updated with fixes.

**Comment by indytechcook** [comment]
This is failing to install in devshop with:

```
Drupal\Component\Plugin\Exception\PluginNotFoundException: The "hms" 
```

Other PR's that failed with that only needed a merge from master but it looks like this one is up to date.  Not sure yet why the error is still throwing.

**Comment by acrollet** [comment]
ok @ethanteague fixes for all previous issues validated :-)

One more thing I found, hopefully the last one!

- Edit node and select a beneficiary term
- Save
- Edit node again and select 'None' for beneficiaries
- Save

The beneficiary term is still saved on the node.

**Comment by ethanteague** [comment]
> ok @ethanteague fixes for all previous issues validated :-)
> 
> One more thing I found, hopefully the last one!
> 
> * Edit node and select a beneficiary term
> * Save
> * Edit node again and select 'None' for beneficiaries
> * Save
> 
> The beneficiary term is still saved on the node.

Good catch, @acrollet ! Updated :+1:

File: docroot/modules/custom/va_gov_backend/js/audience_topics.js [hunk 1]
```diff
@@ -36,43 +50,61 @@
...
 
         if (selection === 'beneficiaries') {
-          // Show beneficiaries field when selected.
+          $('div[id^="edit-field-tags-0-subform-field-non-beneficiares-wrapper"]').hide();
           $('div[id^="edit-field-tags-0-subform-field-audience-beneficiares-wrapper"]').show();
-        } else {
-          // Otherwise clear any selected value in the beneficiaries field and hide it.
-          $('input[id^="edit-field-tags-0-subform-field-audience-beneficiares-none"]').prop('checked', true);
+          // Reset them all.
+          $('fieldset#edit-group-tags input[type="radio"]').attr('checked', false);
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_backend/js/audience_topics.js:84]: It looks like these two lines are run for every condition - can they be moved before the if statement to eliminate the duplicates? More broadly, should we do the hide/resets for all fields at the top and then just show/check whatever is needed in the conditional?

File: docroot/modules/custom/va_gov_backend/js/audience_topics.js [hunk 2]
```diff
@@ -36,43 +50,61 @@
   Drupal.behaviors.vaGovAudienceTopics = {
     attach: function () {
 
+      // Normalize select and taxonomy display on page load.
```
Comment (acrollet) [inline docroot/modules/custom/va_gov_backend/js/audience_topics.js:53]: All test cases are working correctly now, great work @ethanteague !

**Review by acrollet — CHANGES_REQUESTED** [review]

File: docroot/modules/custom/va_gov_backend/js/audience_topics.js (hunk 1 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/js/audience_topics.js]: Was having some head scratchers with things breaking weirdly when I tried to move everything out of a conditional, @acrollet - I did clean it up / make it DRYer, though. Lemme know whatcha think!

Thanks again for your time on this.

Comment (acrollet) [reply docroot/modules/custom/va_gov_backend/js/audience_topics.js]: @ethanteague this looks good, thanks! Easier to see what's going on now. I'd like to review the use cases above one last time on devshop so I'm making a copy of this PR to try and get an environment up: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/3654

Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/js/audience_topics.js]: Much appreciated. @acrollet

**Review by acrollet — APPROVED** [review]
ok @ethanteague this is good to go - I reviewed all previous issues on http://pr3654.ci.cms.va.gov/ :-)

**Comment by ethanteague** [comment]
> ok @ethanteague this is good to go - I reviewed all previous issues on http://pr3654.ci.cms.va.gov/ :-)

Thanks again for working so much on this, @acrollet - really appreciate it!

**Comment by acrollet** [comment]
@ethanteague any time!

**Review by swirtSJW — APPROVED** [review]

**Comment by swirtSJW** [comment]
image

**Pull request merged**